### ISSUE TITLE
Context Creation Problem after Context Creation Failure on CUDA

### ISSUE THREAD

**Issue by Lnaden** [pandegroup/openmm#1875]
There is a subtle bug in how garbage collection is handled after creating a context fails. The symptom is the hardware allocation is not released from what I can tell if you try to create a `Context`and it fails due to unavailable resources. I can only recreate this on the CUDA platform, OpenCL on identical hardware works fine and the cpu's I'm running on have no OpenCL support.

This particular problem is on `Exclusive` mode GPU (one context per GPU). I think this might be related to the problem I am seeing in [OpenMMTools' Travis Test](https://github.com/choderalab/openmmtools/pull/270) where we are intentionally trapping context manipulation failures.

Here is the minimal code to reproduce:
```python
import simtk.openmm as mm
s = mm.System()
s.addParticle(12)
i = lambda: mm.VerletIntegrator(0.002)
p = mm.Platform.getPlatformByName('CUDA')

c = mm.Context(s,i(),p)
try:
    c = mm.Context(s,i(),p)
except Exception:
    pass
del c
c = mm.Context(s,i(),p) # This line will fail.
```
However, this code works fine
```python
import simtk.openmm as mm
s = mm.System()
s.addParticle(12)
i = lambda: mm.VerletIntegrator(0.002)
p = mm.Platform.getPlatformByName('CUDA')
c = mm.Context(s,i(),p)
del c
c = mm.Context(s,i(),p) # This line will work.
```

After the context fails to be created, the output from `nvidia-smi` on the hardware shows no resource consumption nor running processes on the GPU.

The tests I am seeing fail in OpenMMTools may be unrelated, but this is something I found along the way.
cc @jchodera @andrrizzi

**Comment by jchodera** [comment]
Hm. OpenMM's `master` branch has been failing for days too:
https://travis-ci.org/pandegroup/openmm/branches

**Comment by peastman** [comment]
> Hm. OpenMM's master branch has been failing for days too:

I don't think that's related.  It looks like an error trying to upload the documentation.

**Comment by peastman** [comment]
If you don't catch the exception, what is the exception it throws?  That should tell us how far it's getting before it fails.

**Comment by Lnaden** [comment]
The trapped error and the error thrown at the end are the same.
```
Traceback (most recent call last):
  File "cudacreate.py", line 15, in <module>
    c = mm.Context(s,i(),p)
  File "/home/nadenl/miniconda3/envs/ommt/lib/python3.5/site-packages/simtk/openmm/openmm.py", line 11933, in __init__
    this = _openmm.new_Context(*args)
Exception: No compatible CUDA device is available
```

**Comment by peastman** [comment]
I'm trying to reproduce this, but without success.  Possibly that's because I'm using CUDA 9.0 RC.  Here's what happens when I try to put the GPU into exclusive mode:

```
$ sudo nvidia-smi -c 1
Warning: Exclusive_Thread was deprecated! Setting Exclusive_Process instead.
Set compute mode to EXCLUSIVE_PROCESS for GPU 00000000:01:00.0.
All done.
```

In `Exclusive_Process` mode, it's fine for a process to create multiple contexts.  What mode is your GPU in, and what CUDA version are you using?

**Comment by Lnaden** [comment]
GTX1080 GPU 
`Exclusive_process` mode 
Driver v. 367.48 (CUDA 8.0)
OpenMM v. 7.1.1 from Conda

> In Exclusive_Process mode, it's fine for a process to create multiple contexts

Could you elaborate more? I thought if you don't delete the context, trying to create another one tries to start a separate CUDA process which `Exclusive_process` mode on the GPUs does not allow.

I should also mention this is on our cluster and I have to go through a queueing system (LSF) to request access to the GPU. That might not have anything at all to do with the bug, but its a difference.

**Comment by peastman** [comment]
> I should also mention this is on our cluster and I have to go through a queueing system (LSF) to request access to the GPU. That might not have anything at all to do with the bug, but its a difference.

Maybe?  All I can say is that your script works on my computer.  Trying to create the second context does throw an exception (which your script catches).  But once the original context is deleted, it has no problem creating another one.

I've also been looking at the code, and it's not clear to me what could be going wrong.  The obvious explanation is that the constructor successfully creates a context, but then throws an exception later which prevents the destructor from cleaning up the context.  But that shouldn't be the case.  Here's where the exception is thrown from:

https://github.com/pandegroup/openmm/blob/master/platforms/cuda/src/CudaContext.cpp#L211

If you look at the loop immediately above it, you'll see that exception will only be thrown if no call to `cuCtxCreate()` returned `CUDA_SUCCESS`, and hence no context has been created.

**Comment by Lnaden** [comment]
Okay, this is really stupid, its our cluster.

After you said the code should not create the context (and you're right), it gave me the idea to double check the cluster. I tried starting a job, SSH'ing directly into the node, then running the script, which it works fine. It means something with the way our queuing system is set up is not allowing the GPU to fully garbage collect, for.... reasons? Sorry I did not think to check this earlier. Who would have though a job scheduler would screw up a GPU's garbage collection? 

I'll get in touch with out cluster managers and figure out what is going on. If I find anything more, I'll reopen this.

Thanks for looking into it anyways!
